exonum/src/runtime/dispatcher/migration_tests.rs

```diff
@@ -964,25 +1000,108 @@ fn test_migration_commit_with_local_error(local_result: LocalResult) {
+    test_migration_commit_with_local_error(&mut Rig::new(), LocalResult::Saved, "not-good");
 }
 
 #[test]
 #[should_panic(expected = "locally it has finished with an error: This migration is unsuccessful")]
 fn migration_commit_with_local_error_saved_and_node_restart() {
-    test_migration_commit_with_local_error(LocalResult::SavedWithNodeRestart);
+    test_migration_commit_with_local_error(
+        &mut Rig::new(),
+        LocalResult::SavedWithNodeRestart,
+        "not-good",
+    );
+}
+
+#[test]
+fn test_migration_restart() {
+    let artifact_name = "not-good-then-ok";
+    let service_name = "service";
+    let db = Arc::new(TemporaryDB::new());
+
+    // Running migration that should fail.
+    std::panic::catch_unwind(|| {
+        test_migration_commit_with_local_error(
+            &mut Rig::new_with_db(Arc::clone(&db)),
+            LocalResult::Saved,
+            artifact_name,
+        )
+    })
+    .expect_err("Node should panic on unsuccessful migration commit");
+
+    // Check that we have failed result locally.
+    let snapshot = db.snapshot();
+    let schema = DispatcherSchema::new(&snapshot);
+    let res = schema
+        .local_migration_result(service_name)
+        .expect("Schema does not have local result");
+    assert_eq!(res.0.unwrap_err(), "This migration is unsuccessful!");
+
+    // Remove local migration result.
+    let fork = db.fork();
+    remove_local_migration_result(&fork, service_name);
+    db.merge_sync(fork.into_patch())
+        .expect("Failed to merge patch after local migration result remove");
+
+    // Check that local result is removed.
+    let snapshot = db.snapshot();
+    let schema = DispatcherSchema::new(&snapshot);
+    assert!(schema.local_migration_result(service_name).is_none());
+
+    let mut rig = Rig::new_with_db(Arc::clone(&db));
+
+    let fork = rig.blockchain.fork();
+    Dispatcher::commit_migration(&fork, service_name, HashTag::empty_map_hash())
+        .expect("Failed to commit migration");
+    rig.create_block(fork);
+
+    // Check that the migration script has finished.
+    rig.assert_no_migration_threads();
+
+    // Check that local migration result is erased.
+    let snapshot = rig.blockchain.snapshot();
+    let schema = DispatcherSchema::new(&snapshot);
+    let res = schema.local_migration_result(service_name).unwrap();
+    assert_eq!(res.0.unwrap(), HashTag::empty_map_hash());
+
+    // Check current instance migration status.
+    let state = schema.get_instance(100).unwrap();
+    let artifact = ArtifactId::from_raw_parts(
+        MigrationRuntime::ID,
+        artifact_name.to_string(),
+        "0.5.2".parse().unwrap(),
+    );
+    let expected_status = InstanceStatus::migrating(InstanceMigration::from_raw_parts(
+        artifact,
+        Version::new(0, 5, 0),
+        Some(HashTag::empty_map_hash()),
+    ));
+    assert_eq!(state.status, Some(expected_status));
+
+    // Flush the migration.
```

Comment:
Nit: This and the following lines do not test migration restart *per se* and thus could be removed.